ENH: Adding and documenting configs in nx-parallel (#75)
networkx/nx-parallel: Config.md

# Configuring nx-parallel

`nx-parallel` provides flexible parallel computing capabilities, allowing you to control settings like `backend`, `n_jobs`, `verbose`, and more. This can be done through two configuration systems: `joblib` and `NetworkX`. This guide explains how to configure `nx-parallel` using both systems.

> ## 1. Using `joblib`

Reviewer: ```suggestion
## 1. Using `joblib`'s config system
```

> `nx-parallel` relies on [`joblib.Parallel`](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html) for parallel computing. You can adjust its settings through the [`parallel_config`](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html) class provided by `joblib`. For more details, check out the official [joblib documentation](https://joblib.readthedocs.io/en/latest/parallel.html).

Reviewer: ```suggestion
`nx-parallel` relies on [`joblib.Parallel`](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html) for parallel computing. You can adjust its settings through the [`joblib.parallel_config`](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html) class provided by `joblib`. For more details, check out the official [joblib documentation](https://joblib.readthedocs.io/en/latest/parallel.html).
```

### 1.1 Usage

```python
from joblib import parallel_config

# Setting global configs
parallel_config(n_jobs=3, verbose=50)
nx.square_clustering(H)

# Setting configs in a context
with parallel_config(n_jobs=7, verbose=0):
    nx.square_clustering(H)
```

Please refer the [official joblib's documentation](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html) to better understand the config parameters.

Note: Ensure that `nx.config.backends.parallel.active = False` when using `joblib` for configuration, as NetworkX configurations will override `joblib.parallel_config` settings if `active` is `True`.

> ## 2. Using `networkx`

Reviewer: ```suggestion
## 2. Using `networkx`'s backend config system
```

To use NetworkX’s configuration system in `nx-parallel`, you must set the `active` flag (in `nx.config.backends.parallel`) to `True`.

### 2.1 Configs in NetworkX for backends

When you import NetworkX, it automatically sets default configurations for all installed backends, including `nx-parallel`.

```python
import networkx as nx

> print(nx.config)
> ```
> 
> Output:
> 
> ```
> NetworkXConfig(backend_priority=[], backends=Config(parallel=ParallelConfig(active=False, backend='loky', n_jobs=None, verbose=0, temp_folder=None, max_nbytes='1M', mmap_mode='r', prefer=None, require=None, inner_max_num_threads=None, backend_params={})), cache_converted_graphs=True)

Reviewer: The output in the markdown is hard to read. Is there a better way to show this? Here's one attempt. But there is probably a better way -- don't worry about it too much. We can always tweak it later.
```suggestion
print(nx.config.backends.parallel)
```

Output:

```
ParallelConfig(active=False, backend='loky', n_jobs=None, verbose=0, temp_folder=None, max_nbytes='1M', mmap_mode='r', prefer=None, require=None, inner_max_num_threads=None, backend_params={})
```

```

As you can see in the above output, by default, `active` is set to `False`. So, to enable NetworkX configurations for `nx-parallel`, set `active` to `True`. Please refer the [NetworkX's official backend and config docs](https://networkx.org/documentation/latest/reference/backends.html) for more on networkx configuration system.

### 2.2 Usage

```python
# enabling networkx's config for nx-parallel
nx.config.backends.parallel.active = True

# Setting global configs
nxp_config = nx.config.backends.parallel
nxp_config.n_jobs = 3
nxp_config.verbose = 50

nx.square_clustering(H)

# Setting config in a context
with nxp_config(n_jobs=7, verbose=0):
    nx.square_clustering(H)
```

> The configuration parameters are the same as `joblib.parallel_config`, so you can refer to the [official documentation](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html) to better understand these config parameters.

Reviewer: ```suggestion
The configuration parameters are the same as `joblib.parallel_config`, so you can refer to the [official joblib documentation](https://joblib.readthedocs.io/en/latest/generated/joblib.parallel_config.html) to better understand these config parameters.
```

  ```

- **Default Behavior**: By default, `nx-parallel` looks for configs in `joblib.parallel_config` unless `nx.config.backends.parallel.active` is set to `True`.

### 3.3 When Should You Use Which System?

When working with multiple NetworkX backends, it's crucial to ensure compatibility among the backends to avoid conflicts between different configurations. In such cases, using NetworkX's configuration system to configure `nx-parallel` is recommended. This approach helps maintain consistency across backends. For example:

```python
nx.config.backend_priority = ["another_nx_backend", "parallel"]
nx.config.backends.another_nx_backend.config_1 = "xyz"
joblib.parallel_config(n_jobs=7, verbose=50)

nx.square_clustering(G)
```

In this example, if `another_nx_backend` also internally utilizes `joblib.Parallel` (without exposing it to the user) within its implementation of the `square_clustering` algorithm, then the `nx-parallel` configurations set by `joblib.parallel_config` will influence the internal `joblib.Parallel` used by `another_nx_backend`. To prevent unexpected behavior, it is advisable to configure these settings through the NetworkX configuration system.

> If you're using `nx-parallel` independently, either the NetworkX or `joblib` configuration system can be used, depending on your preference.

Reviewer: addressing the question of what "independently" means: one attempt
```suggestion
If your code using `nx-parallel` does not simultaneously use joblib (including inside another networkx backend) then either the `networkx.config` system or the `joblib.parallel_config` system can be used, depending on your preference.
```

Also, I think this paragraph should appear at the top of this section, as I think most users will be in this situation.

Reviewer: Is there any chance of trouble when the other backend doesn't use joblib?
For example, could the config using `joblib.parallel_config` interfere in any way with nx-cugraph or graphblas-algorithms?

I *think* the only way to have a "conflict" is if the other code is using joblib too. Right?

Reviewer: I think it doesn't matter what the codebases of the other backends are: the only config options nx-parallel uses are related to joblib. And the only way for another backend to be influenced by those choices is if they look at or use the config options in joblib. So only backends which use joblib will have a possible conflict over config values. 

In the interest of putting this into place, I am going to merge this PR. And I'll open another one to make further suggestions.

This behavior ensures that `nx-parallel` functions consistently use NetworkX’s settings when enabled, while still allowing Joblib configurations to apply to non-NetworkX parallel tasks.

**Key Takeaway**: When both systems are used together, NetworkX's configuration (`nx.config.backends.parallel`) takes precedence for `nx-parallel` functions. To avoid unexpected behavior, ensure that the `active` setting aligns with your intended configuration system.

### 3.2 Key Differences

- **Parameter Handling**: The main difference is how `backend_params` are passed. Since, in networkx configurations are stored as a [`@dataclass`](https://docs.python.org/3/library/dataclasses.html), we need to pass them as a dictionary, whereas in `joblib.parallel_config` you can just pass them along with the other configurations, as shown below:

  ```py
  nx.config.backends.parallel.backend_params = {"max_nbytes": None}
  joblib.parallel_config(backend="loky", max_nbytes=None)
  ```

- **Default Behavior**: By default, `nx-parallel` looks for configs in `joblib.parallel_config` unless `nx.config.backends.parallel.active` is set to `True`.

### 3.3 When Should You Use Which System?

> - When you're working with multiple NetworkX backends, then to avoid having any conflicts with other backends' code, using networkx to configure nx-parallel would be well suited.

Reviewer: I think you have mentioned potential conflicts before. What are the potential conflicts?

Author: I have updated the `Config.md` with an example of this. Thanks!

- If using `nx-parallel` independently, either system can be used based on your preference.

## 4. Important Considerations

> - **Avoiding Conflicts:** To prevent conflicts, ensure that `networkx.config.backends.parallel.active` is set to `False` when using `joblib`, and `True` when using NetworkX configurations.

Reviewer: Does "when using `joblib`" imply all use of joblib including in nx-parallel and outside nx-parallel? Or does this refer to using `joblib.parallel_config` to set the configuration for nx-parallel?

Author: here, "when using joblib" means when the user is using `joblib.parallel_config` to configure nx-parallel's internal `joblib.Parallel()`.
